paasta_tools/docker_wrapper.py

```diff
@@ -71,31 +76,105 @@ def generate_hostname(fqdn, mesos_task_id):
     return hostname
 
 
-def add_hostname(args, hostname):
-    # Add --hostname argument immediately after 'run' command if it exists
+def add_argument(args, argument):
+    # Add an argument immediately after 'run' command if it exists
     args = list(args)
-
     try:
         run_index = args.index('run')
     except ValueError:
         pass
     else:
-        args.insert(run_index + 1, '--hostname=' + hostname)
-
+        args.insert(run_index + 1, argument)
     return args
 
 
+def get_cpumap():
+    # Return a dict containing the core numbers per physical CPU
+    core = 0
+    cpumap = {}
+    try:
+        with open('/proc/cpuinfo', 'r') as f:
+            for line in f:
+                m = re.match('physical\sid.*(\d)', line)
+                if m:
+                    cpuid = int(m.group(1))
+                    if cpuid not in cpumap:
+                        cpumap[cpuid] = []
+                    cpumap[cpuid].append(core)
+                    core += 1
+    except IOError:
+        pass
+    return cpumap
+
+
+def is_numa_enabled():
+    return os.path.exists('/proc/1/numa_maps')
+
+
+def arg_collision(new_args, current_args):
+    # Returns True if one of the new arguments is already in the
+    # current argument list.
+    cur_arg_keys = []
+    for c in current_args:
+        cur_arg_keys.append(c.split('=')[0])
+    return bool(set(new_args).intersection(set(cur_arg_keys)))
+
+
+def validate_int(i):
+    try:
+        return int(i)
+    except (ValueError, TypeError):
+        return None
+
+
+def validate_float(f):
+    # Ensure we return a float. If input is invalid we return 0.0
+    try:
+        return float(f)
+    except (ValueError, TypeError):
+        return 0.0
+
+
+def generate_cpuset_args(env_args, pinned_numa_node):
+    required_cores = validate_float(env_args.get('MARATHON_APP_RESOURCE_CPUS'))
+    cpumap = get_cpumap()
+    result = []
+
+    # No need to set up docker cpuset options if one of the above fails:
+    if (
+        # NUMA is supported by the system
+        is_numa_enabled() and
```

Comment:
In the new `append_cpuset_args` function I would like better stderr lines to help us debug in the conditions where it isn't adding args:

```
if not numa_enabled():
    log.warning("NUMA optimization was requested but this is not a numa-enabled machine. No optimizations applied")
    return argv
if len(cpus) <= 1:
    log.warning(....
if......

# by now we know we need some args
append...
```
Again these will only show up to stderr when the user has requested numa.

And I dare say that we *should* log to stderr when we a user requests numa and we are going to do something about it.

```
print("NUMA optimization requested. Pinning to NUMA zone %d.", stderr)
```